### ISSUE TITLE
BP mod loader

### ISSUE THREAD

**Issue by justmdj** [UE4SS-RE/RE-UE4SS#597]
### Discussed in https://github.com/UE4SS-RE/RE-UE4SS/discussions/596

<div type='discussions-op-text'>

<sup>Originally posted by **justmdj** July 20, 2024</sup>
Real noob for this tool, I have seen in the documentation where you say the BP mod function of this tool is modeled after Apparate blueprint mod loader. I see there is a video on how to create a BP mod. My question is how do I load a BP mod I have tried adding the BP mod through the BP mod txt file and I have tried to load it using the mod file I am asking for a couple reasons 1 to be able to debug and 2 if I create a BP mod and for some reason have to reset my directory I would like to be able to add the mod back in that I created instead of having to start all over
In Apparate you hit F8 then enter the mod name how do I do this in UE4SS?</div>

**Comment by UE4SS** [comment]
There's no reason to spread the talking around to different places so I removed your discussion, and any discussion can be had in this issue instead.

**Comment by UE4SS** [comment]
If your BP pak file is in `/Content/Paks/LogicMods`, then your mod should automatically be loaded.
You can hit the `INS` key on your keyboard to manually load your mod for testing purposes.

We don't support changing the status of a BP mod via the regular mods.txt file, that file is only for Lua and C++ mods.
There's `/Binaries/Win64/Mods/BPModLoaderMod/load_order.txt`, but that's only for the load order, it doesn't affect whether a mod is loaded or not.
There's also `/Content/Paks/LogicMods/<ModName>/config.lua`, but this file just allows you to change the default mod actor name.

**Comment by justmdj** [comment]
Hi, Thank you for deleting the other I thought I did so after creating the issue from the post. I will give this a try and follow up hopefully I have not messed something up

**Comment by UE4SS** [comment]
Double check that the mod files are in the correct location, and then try loading them manually and then show us the log file.

**Comment by justmdj** [comment]
Ok working on that information for you but Hogwarts updated and now the tool is throwing a fatal error
9:40:17] FName::ToString address: 0x1437f6770 <- Built-in
[09:40:17] Scan attempt 59
[09:40:18] AOB scans could not be completed because of the following reasons:
Was unable to find address for 'StaticConstructObject_Internal'
You can supply your own in 'UE4SS_Signatures/StaticConstructObject'

[09:40:18] Engine Version: 4.27 <- Built-in
[09:40:18] FName::ToString address: 0x1437f6770 <- Built-in
[09:40:18] Fatal Error: AOB scans could not be completed because of the following reasons:
Was unable to find address for 'StaticConstructObject_Internal'
You can supply your own in 'UE4SS_Signatures/StaticConstructObject'

**Comment by UE4SS** [comment]
It would appear that the current release of UE4SS (3.0.1) is no longer compatible with Hogwarts.
You'll need to download an experimental version, like [this one](https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/UE4SS_v3.0.1-137-g6867a08.zip) (current latest) for example.
Apparently, you'll also need manually set the engine version in `UE4SS-settings.ini`, where `Major` should be 4, and `Minor` should be `27`, according to #601.

**Comment by justmdj** [comment]
T/Y doing that now

**Comment by justmdj** [comment]
Hi,
Ok I tried what you outline above and as soon as i click a button to go from the load screen to the screen where you select your Hogwarts game you want to ply IE character name It crashed to desktop i also went in and tried to disable the bUseUObjectArrayCache = true to false. still crashes to desktop

**Comment by Noodneek** [comment]
> It would appear that the current release of UE4SS (3.0.1) is no longer compatible with Hogwarts. You'll need to download an experimental version, like [this one](https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/UE4SS_v3.0.1-137-g6867a08.zip) (current latest) for example. Apparently, you'll also need manually set the engine version in `UE4SS-settings.ini`, where `Major` should be 4, and `Minor` should be `27`, according to #601.

> Hi, Ok I tried what you outline above and as soon as i click a button to go from the load screen to the screen where you select your Hogwarts game you want to ply IE character name It crashed to desktop i also went in and tried to disable the bUseUObjectArrayCache = true to false. still crashes to desktop

I use the current release UE4SS 3.0.1, not the experimental version, and with edited `UE4SS-settings.ini` to set [EngineVersionOverride] to 4.27 it worked flawlessly until about two days ago, when it started to cause crashes for some reason...
Either it crushes to desktop when loading, just before "preparing shaders" ([here](https://pastebin.com/P3ZDcqVF)'s WinDbg analysis of the crash.dmp) or in a middle of a broom flight ([here](https://pastebin.com/7MW2kTU2)'s the analysis of the crash.dmp). If the game loads (which is about 50/50 chance), it usually crashes after several min (the longest I played before it crashed was about an hour)... And when I load saves or fast travel, with the UE4SS-RE it is much slower and can also cause a crash...

**Comment by Noodneek** [comment]
Just tried [the experimental version](https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/UE4SS_v3.0.1-137-g6867a08.zip) with edited UE4SS-settings.ini to set [EngineVersionOverride] to 4.27, so far so good... The only external mod I use is  [Better and instant timeskip](https://www.nexusmods.com/hogwartslegacy/mods/1158?tab=description&BH=0)...

**Comment by justmdj** [comment]
@Noodneek yes it has worked for me except for a couple mods and i was working on getting thos working then the patch hit and broke everything lol

**Comment by UE4SS** [comment]
It seems like it's working for @Noodneek, but still isn't working for you @justmdj ?
That's weird if that's the case.
Try disabling all your mods if you didn't already, just to verify that this problem isn't related to a mod.

**Comment by Noodneek** [comment]
> @Noodneek yes it has worked for me except for a couple mods and i was working on getting thos working then the patch hit and broke everything lol

@justmdj, July 24 patch broke it for me as well - read my post here: https://github.com/UE4SS-RE/RE-UE4SS/issues/601.
V.2.5.1 was unable to find address for 'StaticConstructObject_Internal', exactly as you wrote; then I upgraded to v.3.0.1, but it failed to find EngineVersion, which I resolved by editing `UE4SS-settings.ini` to set [EngineVersionOverride] to `Major = 4` and `Minor = 27`. It worked fine but after 2 days started causing frequent crashes.
[This version](https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/UE4SS_v3.0.1-137-g6867a08.zip) works fine though, no crashes or lags so far... (Big thanks to @UE4SS for suggesting it!). So I also think your particular modes could be the cause of your crashes, and not the UE4SS-RE...
[!] Pay attention that all the files of the [new experimental version](https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/UE4SS_v3.0.1-137-g6867a08.zip) (except for `dwmapi.dll`) are now in a separate folder `ue4ss`, and not directly in the `\Phoenix\Binaries\Win64` folder!

**Comment by UE4SS** [comment]
> Pay attention that all the files of the [new experimental version](https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/UE4SS_v3.0.1-137-g6867a08.zip) (except for `dwmapi.dll`) are now in a separate folder `ue4ss`, and not directly in the `\Phoenix\Binaries\Win64` folder!

This is true, but it's still compatible with the old file structure as well but obviously then you need to extract the contents of the ue4ss directory instead of the ue4ss directory itself.
Choose either one, but we obviously are going to recommend the new structure since that's the one we're going to go forward with and the old one will be deprecated and might be deleted in a future version after everyone has caught up to the new one.

**Comment by justmdj** [comment]
@UE4SS @Noodneek I figured out what was causing the Crash to desktop I had thought about it being mobs as well but found instead that it was me I tend to make backups of everything so what i did when I installed this version was remove win64 instead of just the files and copied my old win64 to the directory then installed this version and changed the version number so in essence it was this version of UE4SS with 4 27 as engine with an exe file 4 23 so yeah my silliness caused it. that said it is loading and working correctly but when I enable the debug tools the dos prompt comes up with feedback like normal but the debugging tools window is a blank white screen with nothing if either of you have thoughts on why this might be or how to fix it I would appreciate the advice

**Comment by UE4SS** [comment]
> @UE4SS @Noodneek I figured out what was causing the Crash to desktop I had thought about it being mobs as well but found instead that it was me I tend to make backups of everything so what i did when I installed this version was remove win64 instead of just the files and copied my old win64 to the directory then installed this version and changed the version number so in essence it was this version of UE4SS with 4 27 as engine with an exe file 4 23 so yeah my silliness caused it. that said it is loading and working correctly but when I enable the debug tools the dos prompt comes up with feedback like normal but the debugging tools window is a blank white screen with nothing if either of you have thoughts on why this might be or how to fix it I would appreciate the advice

Try changing `GraphicsAPI` in UE4SS-settings.ini to `dx11`.

**Comment by Noodneek** [comment]
> > @UE4SS @Noodneek I figured out what was causing the Crash to desktop I had thought about it being mobs as well but found instead that it was me I tend to make backups of everything so what i did when I installed this version was remove win64 instead of just the files and copied my old win64 to the directory then installed this version and changed the version number so in essence it was this version of UE4SS with 4 27 as engine with an exe file 4 23 so yeah my silliness caused it. that said it is loading and working correctly but when I enable the debug tools the dos prompt comes up with feedback like normal but the debugging tools window is a blank white screen with nothing if either of you have thoughts on why this might be or how to fix it I would appreciate the advice
> 
> Try changing `GraphicsAPI` in UE4SS-settings.ini to `dx11`.

Yeah, I was wondering about that too. Debugging tools window stopped appearing after I swapped v.2.5.1 for v.3.0.1. I changed `GuiConsoleVisible` from 0 to 1, it appeared, but was a blank white screen. Changing GraphicsAPI in `UE4SS-settings.ini` from `opengl` to `dx11` fixed it, thanks!

**Comment by justmdj** [comment]
dx11 solved the gui issue now I can get back to the issue with the mod loading..... maybe i will get lucky and it will just load now

**Comment by zigonsehpaaaa** [comment]
I put pack folder at correct place nad ue4 runs properly but when I open up BPMod loader menu no mod I installed shows up??
